src/pybind/mgr/nfs/cluster.py

```diff
@@ -73,11 +74,30 @@ def _call_orch_apply_nfs(
             ispec = IngressSpec(service_type='ingress',
                                 service_id='nfs.' + cluster_id,
                                 backend_service='nfs.' + cluster_id,
+                                placement=PlacementSpec.from_string(placement),
                                 frontend_port=port,
                                 monitor_port=7000 + port,   # semi-arbitrary, fix me someday
                                 virtual_ip=virtual_ip)
             completion = self.mgr.apply_ingress(ispec)
             orchestrator.raise_if_exception(completion)
+        elif virtual_ip:
+            # nfs + keepalived
```

Comment:
I just happend to be skimming this and this part jumped out at me as looking like a fair bit of duplication and it's difficult for me to see what differs between these blocks.  My suggestion would be to move the duplicated parts into a function call. Then for the parts that differ pass specific arguments. example:

```python
if virtual_ip and not keepalive_only:
    nfs_spec, ingress_spec = self.build_specs(ingress=True, nfs_virtual_ip=None, ingress_virtual_ip=virtual_ip)
elif virtual_ip:
    nfs_spec, ingress_spec = self.build_specs(ingress=True, nfs_virtual_ip=virtual_ip, ingress_virtual_ip=None)
else:
   nfs_spec, ingress_spec = self.build_specs(ingress=False)
orchestrator.raise_if_exception(self.mgr.apply_nfs(nfs_spec))
if ingress_spec:
    orchestrator.raise_if_exception(self.mgr.apply_ingress(ingress_spec))
```

Something along those lines.